fixed rcl_wait return error when timer cancelled

ladianchad opened the pull request:
when timer canclled, `rcl_timer_get_time_until_next_call` return `RCL_RET_TIMER_CANCELED`
rcl_wait return with "error not set" like below ( we checked `rcl_timer_is_canceled` but timer is atomic, we can miss this)

```bash
[planner_server-3] terminate called after throwing an instance of 'rclcpp::exceptions::RCLError'
[planner_server-3]   what():  rcl_wait() failed: error not set
```
so, I changed check timer canceled by `rcl_timer_get_time_until_next_call`'s return value.

Would these solutions be appropriate?

ladianchad commented:
https://github.com/ros2/rclcpp/issues/1977
this bug also mentioned at this issue

fujitatomoya reviewed (approved):
lgtm!

fujitatomoya commented:
@ladianchad thanks for the contribution, this looks good to me. can you address DCO error?

fujitatomoya commented:
@iuhilnehc-ynos requesting another review on this.

CC: @ivanpauno

ladianchad commented:
> @ladianchad thanks for the contribution, this looks good to me. can you address DCO error?

yes, I fixed DCO error and rebased my unnecessary commits

iuhilnehc-ynos commented:
LGTM. The default branch to push is `rolling`, then we can backport it to `humbel`.

ladianchad commented:
oh. thanks i changed pr branch to humble

ivanpauno commented on rcl/src/rcl/wait.c:
```diff
@@ -558,19 +558,15 @@ rcl_wait(rcl_wait_set_t * wait_set, int64_t timeout)
       // TODO(sloretz) fix spurious wake-ups on ROS_TIME timers with ROS_TIME enabled
       int64_t timer_timeout = INT64_MAX;
-      ret = rcl_timer_get_time_until_next_call(wait_set->timers[i], &timer_timeout);
+
```
```suggestion
```

ivanpauno reviewed (approved):
LGTM!

fujitatomoya commented:
@ladianchad can you address @ivanpauno comment ? then i will start CI.

ladianchad commented:
> @ladianchad can you address @ivanpauno comment ? then i will start CI.

I applied review in [1554188](https://github.com/ros2/rcl/pull/1003/commits/1554188cf02e0f0b640d0e4ceff61c96773cea6f) :)

fujitatomoya commented:
@ladianchad thanks for the quick response.

CI:
* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=17174)](http://ci.ros2.org/job/ci_linux/17174/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=11717)](http://ci.ros2.org/job/ci_linux-aarch64/11717/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=17598)](http://ci.ros2.org/job/ci_windows/17598/)

ladianchad commented:
Linux Build Status is unstable, is it ok?? I don't quite understand ROS CI yet.

fujitatomoya commented:
I do not think that is related.

CI(Retry Linux):

* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=17176)](http://ci.ros2.org/job/ci_linux/17176/)

ladianchad commented:
Thanks!

fujitatomoya commented:
@Mergifyio backport humble galactic foxy

ErikOrjehag commented:
Thank you for fixing this one! :D
